gc.c

```diff
@@ -8258,6 +8058,164 @@ gc_marks_step(rb_objspace_t *objspace, size_t slots)
 }
 #endif
 
+static bool
+gc_compact_heap_cursors_met_p(rb_heap_t *heap)
+{
+    if (heap->sweeping_page && heap->compact_cursor) {
+        return heap->sweeping_page == heap->compact_cursor;
+    }
+    else {
+        return false;
+    }
+}
+
+static bool
+gc_compact_move(rb_objspace_t *objspace, rb_heap_t *heap, VALUE src)
+{
+    GC_ASSERT(BUILTIN_TYPE(src) != T_MOVED);
+    rb_heap_t *dheap = heap;
+
+    if (gc_compact_heap_cursors_met_p(dheap)) {
+        return false;
+    }
+    while (!try_move(objspace, dheap, dheap->sweeping_page, src)) {
+        if (dheap->sweeping_page->flags.before_sweep == TRUE) {
+            struct gc_sweep_context ctx = {
+                .page = dheap->sweeping_page,
+                .final_slots = 0,
+                .freed_slots = 0,
+                .empty_slots = 0,
+            };
+            gc_sweep_page(objspace, heap, &ctx);
+        }
+
+        if (try_move(objspace, dheap, dheap->sweeping_page, src)) {
+            break;
+        }
+
+        dheap->sweeping_page = ccan_list_next(&dheap->pages, dheap->sweeping_page, page_node);
+        if (gc_compact_heap_cursors_met_p(dheap)) {
+            return false;
+        }
```

Comment:
Maybe we can do something like this? And then `try_move` can pull slots on pages from the `dheap->free_pages` rather than from `dheap->sweeping_page`. 

```suggestion
        if (gc_compact_heap_cursors_met_p(dheap)) {
            return false;
        }

        struct gc_sweep_context ctx = {
            .page = dheap->sweeping_page,
            .final_slots = 0,
            .freed_slots = 0,
            .empty_slots = 0,
        };
        gc_sweep_page(objspace, heap, &ctx);
        heap_add_freepage(dheap, dheap->sweeping_page);

        dheap->sweeping_page = ccan_list_next(&dheap->pages, dheap->free_pages, page_node);
```

Reply:
We could do that. I think we'd need to null out `free_pages` at the beginning of compaction (and then rebuild it again at the end) to guarantee the correct ordering of the `free_pages` and that there are no cycles in the list, otherwise we could be compacting inconsistently.

I'm not entirely sure why checking for `before_sweep` here is a huge concern given that this is a fairly tight loop that only runs during compaction. I understand that the `sweeping_pages` list should be a queue of unswept pages, but during compaction I'm really using it as a proxy for the scan cursor. Compaction is not incremental, it happens during a major GC, and in a fairly tight loop and so `sweeping_pages` can't be left in an inconsistent state when the interpreter is running user code.

Reply:
> `gc_sweep_start_heap` already nulls out the `free_pages` of the heap.

👍 You're right. I'd missed that - thanks for the catch.

> I think it should be an invariant that `before_sweep` is true for all pages on the `sweeping_pages` linked list.

Yup, this does make sense to me. I will refactor to use `free_pages` as an intermediary. 

> If we don't move `sweeping_page` forward, wouldn't sweeping re-sweep the last swept page during sweeping right after compaction has completed?

In the current implementation, yes. However compaction uses `gc_sweep_page` directly, rather than `gc_sweep_step` - so whilst they have been swept twice, the actual sweeping process will be a no-op the second time around, and then `gc_sweep_step` will do the required accounting and heap manipulation it needs to.

In https://github.com/ruby/ruby/pull/5691 I'm starting to work on making it easier to use `gc_sweep_page` from outside of `gc_sweep_step` with the intention of removing the need for the second sweep entirely.

Thanks again for the help and discussion Peter 🙇